### ISSUE TITLE
Torch multiprocessing probelms

### ISSUE THREAD

**Issue by pattonw** [funkey/gunpowder#100]
What do I want? The ability to use a pipeline that looks something like:
``` python
pipeline = gunpowder.torch.nodes.Predict(...) + gunpowder.torch.nodes.Scan(...)
```
What goes wrong:
First error: `RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method`

Attempted to solve with using `multiprocessing.set_start_method("spawn")`
Second error: `RuntimeError: context has already been set`
I think this is the first problem that needs fixing: importing gunpowder should not set the spawn method before a user gets the chance to determine what the spawn method should be. I've tracked this down to the class variable `__next_id` created in the class `Batch` in file "gunpowder/batch.py". It seems this line sets the spawn method to "fork" if no method has been set yet.
Additionally the `LocalServer` in "gunpowder/tensorflow/local_server.py" also has a class variable that leads to the spawn method being set on import.

Further problems:
After removing the two mentioned variables that cause Multiprocessing to set a default spawn method, you can set the spawn method with `multiprocessing.set_spawn_method("spawn")`. You then get an error: `AssertionError: can only join a started process`. This seems to be getting thrown in "gunpowder/producer_pool.py" on line 76 at `self.__watch_dog.join()`. It seems to happen

Does this mean we can't use double underscore methods if we want to be multiprocessing friendly?

Continuing:
After removing all "__{name}" methods and replacing them with "_{name}" in `ProducerPool`, the precache.py tests still seem to hang forever. This seems to be an issue with logging errors in tests, so I created some simple scripts to do the same thing outside of a testing framework. Now I get the following error:
`RuntimeError: Attempted to send CUDA tensor received from another process; this is not currently supported. Consider cloning before sending.`
The stack trace here is particularly unhelpful so I am having a lot of difficulty figuring out where this is coming from.

See branch `torch-multiprocessing` for changes I have made to explore this problem.

**Comment by funkey** [comment]
The problem you are running into actually stems from `torch.Predict`. In its constructor, it is already initializing CUDA (by sending the model to the "cuda" device, if one exists). This only works if `torch.Predict` is used in a single process. `Scan`, however, does spawn processes (if `num_workers` is larger than 1), which now already have CUDA initialized, leading to the error you observe.

One solution¹ is to defer initialization of anything that relates to CUDA to the process in which it is used. `gunpowder`s predict nodes (`torch.Predict` and `tensorflow.Predict`) inherit from `GenericPredict`, which contains a [`start()`](https://github.com/funkey/gunpowder/blob/master/gunpowder/nodes/generic_predict.py#L172-L179) method for exactly this purpose. Its documentation says:

> This method will be called before the first call to `predict`, from the same process that `predict` will be called from. Use this to initialize your model and hardware.

I fixed the implementation of `torch.Predict` to initialize CUDA in `start()`, see branch [torch_predict_multiprocessing](https://github.com/funkey/gunpowder/tree/torch_predict_multiprocessing). With this change, the following example script runs fine:

```python
import gunpowder as gp
import numpy as np
import torch


class ArraySource(gp.BatchProvider):

    def __init__(self, array_key, array):
        self.array_key = array_key
        self.array = array

    def setup(self):
        self.provides(self.array_key, self.array.spec)

    def provide(self, request):

        batch = gp.Batch()
        batch[self.array_key] = self.array.crop(request[self.array_key].roi)
        return batch

if __name__ == '__main__':

    raw = gp.ArrayKey('RAW')
    pred = gp.ArrayKey('PRED')

    raw_array = gp.Array(
        np.ones((100,), dtype=np.float32)*10,
        gp.ArraySpec(roi=gp.Roi((0,), (100,)), voxel_size=(1,)))

    model = torch.nn.Sigmoid()
    model.eval()

    scan_request = gp.BatchRequest()
    scan_request[raw] = gp.Roi((0,), (10,))
    scan_request[pred] = gp.Roi((0,), (10,))

    request = gp.BatchRequest()
    request[raw] = gp.Roi((0,), (100,))
    request[pred] = gp.Roi((0,), (100,))

    pipeline = (
        ArraySource(raw, raw_array) +
        gp.torch.Predict(model, {'input': raw}, {0: pred}, device='cuda') +
        gp.Scan(scan_request, num_workers=10))

    with gp.build(pipeline):
        batch = pipeline.request_batch(request)

    print(batch[raw].data)
    print(batch[pred].data)
```

This was tested with the following packages:
```
cudatoolkit               10.0.130                      0  
cudnn                     7.6.5                cuda10.0_0
python                    3.7.7                hcff3b4d_5  
pytorch                   1.3.1           cuda100py37h53c1284_0  
pytorch-gpu               1.3.1                         0  
```

Let me know if this fix works for your case.

¹: The other solution is to use the `spawn` start method, as the error message suggests. Unfortunately, the error message fails to mention that the better solution is to avoid having CUDA shared between processes in the first place, if possible.

> I think this is the first problem that needs fixing: importing gunpowder should not set the spawn method before a user gets the chance to determine what the spawn method should be.

I think in this case, if a user really wants to change the start method, it should be set before importing `gunpowder`:
```python
import multiprocessing
multiprocessing.set_start_method('spawn')
import gunpowder
```

> Does this mean we can't use double underscore methods if we want to be multiprocessing friendly?

No, `__` methods have nothing to do with multiprocessing. They are just Python's version of "private" functions that can not be overwritten by subclasses. `__init__` and `__repr__`, for example, use this concept. See the [documentation](https://docs.python.org/3/tutorial/classes.html#tut-private) for more details.

**Comment by funkey** [comment]
As discussed off-site, this was indeed the problem. Fixed in 3e9e1dad09b7d95c8bad25e9bf4986b24a584811, together with a similar issue in `torch.Train`.
